File: Examples/Example-ObjC/Example-ObjC/ViewController.m
//
//  ViewController.m
//  Example-ObjC
//
//  Created by Юрий Буянов on 17/09/2018.
//  Copyright © 2018 Example Ltd. All rights reserved.
//

#import "ViewController.h"
#import <StoreKit/StoreKit.h>
#import <Apptilaus/Apptilaus-umbrella.h>

@interface ViewController ()<SKProductsRequestDelegate, SKPaymentTransactionObserver>

@property (nonatomic) IBOutletCollection(UIButton) NSArray<UIButton *> *buttons;
@property (nonatomic) NSArray<SKProduct *> *products;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [[SKPaymentQueue defaultQueue] addTransactionObserver:self];
    
    NSArray<NSString *> *productIds = @[@"test.purchase.one",
                                        @"test.subscription.renew.week",
                                        @"test.subscription.norenew.week"];
    
    SKProductsRequest *req = [[SKProductsRequest alloc] initWithProductIdentifiers:[NSSet setWithArray:productIds]];
    req.delegate = self;
    
    [req start];
}

- (IBAction)buttonTap:(UIButton *)sender {
    NSUInteger index = [self.buttons indexOfObject:sender];
    if (index != NSNotFound && index < self.products.count) {
        SKProduct *product = self.products[index];
        SKPayment *payment = [SKPayment paymentWithProduct:product];
        [[SKPaymentQueue defaultQueue] addPayment:payment];
        
    }
}


#pragma mark - SKProductsRequestDelegate

- (void)productsRequest:(SKProductsRequest *)request didReceiveResponse:(SKProductsResponse *)response {
    NSNumberFormatter *formatter = [[NSNumberFormatter alloc] init];
    formatter.numberStyle = NSNumberFormatterCurrencyStyle;
    
    for (NSUInteger i = 0; i < MIN(response.products.count, self.buttons.count); i++) {
        SKProduct *product = response.products[i];
        formatter.locale = product.priceLocale;
        NSString *title = [NSString stringWithFormat:@"%@ %@", product.localizedTitle, [formatter stringFromNumber:product.price]];
        [self.buttons[i] setTitle:title forState:UIControlStateNormal];
    }
    
    self.products = [response.products copy];
}

#pragma mark - SKPaymentTransactionObserver

- (void)paymentQueue:(SKPaymentQueue *)queue updatedTransactions:(NSArray<SKPaymentTransaction *> *)transactions {
    for (SKPaymentTransaction *transaction in transactions) {
        if (transaction.transactionState == SKPaymentTransactionStatePurchased) {
            NSDictionary<NSString *, id> *customParams = @{
                 @"foo": @"bar",
                 @"goats_count": @123
            };
            [[ApptilausManager shared] registerWithTransaction:transaction customParams:customParams];
            [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
        } else if (transaction.transactionState == SKPaymentTransactionStateFailed) {
            NSLog(@"Transaction %@ failed with error: %@", transaction, transaction.error);
        }
    }
}

@end

File: Examples/Example-Swift/Example-Swift/ViewController.swift
import UIKit
import StoreKit
import Apptilaus

class ViewController: UIViewController {
    
    private let productIds = ["test.purchase.one",
                              "test.subscription.renew.week",
                              "test.subscription.norenew.week"]
    
    private var buttons: [UIButton] = []
    
    private var products: [String : SKProduct] = [:]
    
    @IBOutlet private var purchase1: UIButton?
    @IBOutlet private var purchase2: UIButton?
    @IBOutlet private var purchase3: UIButton?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        SKPaymentQueue.default().add(self)
        
        self.buttons = [self.purchase1!, self.purchase2!, self.purchase3!]
        let req = SKProductsRequest(productIdentifiers: Set(self.productIds))
        req.delegate = self;
        req.start()
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func buttonTap(sender: UIButton) {
        if let index = self.buttons.index(of: sender),
            let product = self.products[self.productIds[index]] {
            let payment = SKPayment(product: product)
            SKPaymentQueue.default().add(payment)
        }
        
    }
    
}

extension ViewController: SKProductsRequestDelegate {
    func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        for product in response.products {
            self.products[product.productIdentifier] = product
            if let index = self.productIds.index(of: product.productIdentifier) {
                let button = self.buttons[index]
                formatter.locale = product.priceLocale
                let title = "\(product.localizedTitle) \(String(describing: formatter.string(from: product.price)))"
                button.setTitle(title, for: .normal)
            }
        }
        print("response: \(response.products)")
    }
}

extension ViewController: SKPaymentTransactionObserver {
    func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
        for transaction in transactions {
            print ("updated tx: \(String(describing: transaction.transactionIdentifier)) state: \(transaction.transactionState))")
            if transaction.transactionState == .purchased {
                let customParameters: [String : Any] = [
                    "foo": "bar",
                    "goats_count": 123
                ];
                
                ApptilausManager.shared.register(transaction: transaction, customParams: customParameters)

                SKPaymentQueue.default().finishTransaction(transaction)
            } else if transaction.transactionState == .failed,
                let error = transaction.error {
                print ("tx: \(String(describing: transaction.transactionIdentifier)) failed with error: \(error)")
            }
        }
    }
    
}
